dxFilterBuilder: Add custom operations and the "between" operation

komarovalexander opened the pull request:
```
customOperations: [{           
    name: "myoperationwithdates",
    caption: "custom operation with date",
    dataTypes: ["date"], 
    icon: "range",
    calculateFilterExpression: function(filterValue, field) {
        var date = new Date(); 
        date.setDate(filterValue.getDate() - 7); 
        return [field.dataField, “>”, date];
    },
    editorTemplate: function(conditionInfo, container) { },
    customizeText: function(e) { },
    hasValue: true 
}]
```
settings for the "between" were added by default, add { name: "between" } in customOperations to enable it: 
```
customOperations: [{           
    name: "between"   
}]
```
valid filter expression of custom operations for the data source can be obtained through the getFilterExpression() method:
`
dataSource.filter(filterBuilder.getFilterExpression())
`

<!--
Make sure that you've read the "Contributing Code and Content" section in CONTRIBUTING.md

If you are submitting a bug fix, the subject should contain "fixes ISSUE_ID" or "resolves ISSUE_ID".

In addition, please clarify:

1. What did you do? 
2. How did you do it?
3. How should we verify this?

-->

komarovalexander committed:
dxFilterBuilder: + getFilterExpression()

komarovalexander committed:
dxFilterBuilder: + customOperations

komarovalexander committed:
Fix optionChanged_bundled.tests.js

komarovalexander committed:
dxFilterBuilder: +customOperation.editorOptions &  +customOperation.customizeText

komarovalexander committed:
dxFilterBuilder: Fix doccoments

komarovalexander committed:
dxFilterBuilder:
1) remove editorOptions;
2) finish customOperations

komarovalexander committed:
dxFilterBuilder:
1) Add between
2) Add filterOperation to editorPreparing events

komarovalexander committed:
dxFilterBuilder: Fix tests for nojquery

komarovalexander committed:
dxFilterBuilder: Fix empty value for between

komarovalexander committed:
dxFilterBuilder: Fix customizeText in between

komarovalexander committed:
dxFilterBuilder: Fix doc coments for calculateFilterExpression

vconst commented on js/ui/filter_builder/utils.js:
```diff
@@ -319,6 +355,47 @@ function getNormalizedFields(fields) {
+    }
+
+    var criteria = getGroupCriteria(value),
+        getConditionFilterExpression = function(condition) {
```
Extract this method from closure. A separate method will be faster and the getFilterExpression will be smaller in this case.

vconst commented on js/ui/filter_builder/utils.js:
```diff
@@ -524,6 +620,47 @@ function setFocusToBody() {
     }
 }
 
+function getBetweenConfig(editorTemplate) {
```
The between operation implementation (config and editor) may be moved to separate file.

komarovalexander committed:
dxFilterBuilder: ignore customOperation without dataType

komarovalexander committed:
dxFilterBuilder: Fix after review

komarovalexander committed:
dxFilterBuilder: Fix between editor width

komarovalexander committed:
dxFilterBuilder: Fix customize text with lookup

komarovalexander committed:
Merge remote-tracking branch 'DevExtreme/18_1' into between

# Conflicts:
#	js/ui/shared/ui.editor_factory_mixin.js

komarovalexander committed:
dxFilterBuilder: getFilterExpression returns null if calculateFilterExpression returns nothing

komarovalexander committed:
dxFilterBuilder:
1) getFilterExpression() always returns filter with group value
2) Fix between calculateFilterExpression for value = [null, null]

komarovalexander committed:
dxFilterBuilder: Add empty icon for custom operation

komarovalexander committed:
Rename empty icon class from context-menu-sort-none to context-menu-icon-none

komarovalexander committed:
Rename empty icon class from context-menu-icon-none to icon-none
